teps/draft-20200715-access-control-via-sa.md

```diff
@@ -0,0 +1,249 @@
+Tasks) for specific users or groups of users.    
+
+As part of both
+
+- Continuing the perception Tekton as a true, Kubernetes native CI/CD system, that leverages all the inherent 
+capabilities of Kubernetes
+- Reaching at least equal capability, compared to existing options, across the spectrum of concerns of a CI/CD system
+
+This proposal articulates a path to finer grained authorization/access control around the various Tekton concepts.
+
+### Goals
+
+One goal of this TEP is to define an approach for access control for references from one Tekton object to another in
+a manner that is natural for Kubernetes cluster administrators.
+
+A second goal of this TEP is introduce such access control in a manner that does not immediately disrupt existing 
+deployments of Tekton, given the free access to all approach that currently exists for these Tekton object to Tekton
+object references.
+
+### Non-Goals
+
+Other valid authorization related scenarios have been raised in various community discussions.  This proposal 
+does not attempt to address all of them.
+
+Among these various scenarios, there has also been discussion around introducing OPA (either via the admission webhook 
+for Kubernetes provided by the OPA org, or some other form of integration) into Tekton.  This proposal is not taking
+on how OPA would be integrated / leveraged by Tekton. 
+
+## Requirements
+
+Kubernetes administrators have means to enact control over which ClusterTasks, Tasks, or Pipelines are referenced
+by a PipelineRun or TaskRun.
+
+Kubernetes administrators have a means to opt into such enforcement on a granular basis (per namespace). 
+
+## Proposal
+
+An addition of a Kubernetes Validating Admission Webhook to the current Tekton Webhooks Deployment can 
+serve as the contact point for intercepting any create/update requests for TaskRun and PipelineRun and 
+then see if the ServiceAccount specified (or "default" if non specified) has permission to get/read any
+referenced ClusterTask/Task/Pipeline objects.
+
+A Kubernetes RBAC Subject Access Review (SAR) check will be executed to see if the ServiceAccount in question
+is part of a Role/RoleBinding or ClusterRole/ClusterRoleBinding that allows access to the referenced 
+ClusterTask, Task, or PipelineTask.  If access is denied, the admission webhook will prevent creation/update
+from completing per the contract specified by the Kubernetes Validating Admission Webhook API.
+
+By default, this new webhook will not perform any checks.  However, on each request it will interrogate 
+a documented environment variable (which is specified on the Tekton Webhook Deployment).  If with this environment
+variable:
+
+- "all" is specified, preceded/followed by a space, factoring in beginning and end of the value of the environment 
+variable
+- or a space delimited list of names corresponding to namespaces is provide (where "all" could be one of those 
+entries as well, hence turning the feature on for every namespace)
```

Comment:
What would you think about creating a separate, initially experimental, project around a Tekton admission webhook for apply security controls? We could allow users to integrate that webhook with systems like OPA.

This way the existing Tekton controller and webhook can be concerned with just pipeline execution, and if users want to enforce their custom security requirements, they can install an optional controller to do so.

Reply:
OK that is the clarification I asked for in https://github.com/tektoncd/community/pull/152#issuecomment-664476168 :-) ... thanks

Certainly that is possible.  The pros/cons are pretty typical .. at least some of them
- pro: separate of concerns
- pro: no on/off switch for tektoncd/pipelines 
- con: more work :-)
- con: some more pod sprawl
- it being separate implies off by default, where in these discussion we talked about on by default ... but when also considering the moving parts around OCI and whether Task/ClusterTask/Pipeline get replaced also muddies those waters.

But I could see how one might prefer those pros in relation to the cons.

That said, enablement of enforcement on a per namespace basis I think would still be useful config.  It corresponds to what sig-auth folks have noted on the problem with PSP being an all or nothing thing.  Existing tekton users could benefit from enabling on a per namespace and migrating in a more granular fashion.